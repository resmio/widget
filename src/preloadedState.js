import moment from 'moment'
// import availabilities from './data/availabilities'

const preloadedState = {
  availabilities: [],
  BookingId: '',
  guestEmail: '',
  guestName: '',
  guestPhone: '',
  newsletterSubscription: false,
  maxGuests: 6,
  minGuests: 1,
  selectedDate: moment(),
  selectedGuests: 2,
  selectedAvailability: '',
  status: 'confirmed',
  timePeriodSelected: 0,
  timeFocused: '',
  buttonColor: '#3E4862',
  // Probably don't need the defaultHeight
  defaultHeight: '380px', // = 500px - 60px header - 60px footer
  defaultWidth: '330px',
  facility: 'meson-california-2',
  headerColor: '#53628C',
  headerImage: 'https://zenezake.files.wordpress.com/2015/07/img_6715.jpg',
  headerTextColor: 'white',
  numberOfPanels: 2,
  renderAtMaxSize: false,
  currentPanel: 2,
  guestSelectorState: 'collapsed',
  dateSelectorState: 'collapsed',
  timeSelectorState  : 'expanded',
  availabilitiesFetching: false,
  availabilitiesError: false
}

export default preloadedState
