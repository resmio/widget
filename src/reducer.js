function reducer(state = {}, action) {

  switch (action.type) {

    case 'GUEST_REMOVE':
      if (state.selectedGuests > state.minGuests) {
        return Object.assign({}, state, {
          selectedGuests: state.selectedGuests - 1
        })
      } else {
        return state
      }

    case 'GUEST_ADD':
      if (state.selectedGuests < state.maxGuests) {
        return Object.assign({}, state, {
          selectedGuests: state.selectedGuests + 1
        })
      } else {
        return state
      }

    case 'UI_GUEST_DROPDOWN_OPEN':
      return Object.assign({}, state, {
        guestSelectorCollapsed: false
      })

    case 'GUEST_SELECT':
      return Object.assign({}, state, {
        selectedGuests: parseInt(action.payload, 10) + 1,
        guestSelectorCollapsed: true
      })

    case 'UI_CALENDAR_SWITCH_FOCUS':
      return Object.assign({}, state, {
        calendarFocused: !state.calendarFocused
      })

    case 'DATE_SELECT':
      return Object.assign({}, state, {
        selectedDate: action.payload
      })

    case 'UI_PANEL_ADVANCE':
      if (state.currentPanel < state.numberOfPanels) {
        return Object.assign({}, state, {
          currentPanel: state.currentPanel + 1
        })
      } else {
        return state
      }

    case 'UI_PANEL_REDUCE':
      if (state.currentPanel >= 2) {
        return Object.assign({}, state, {
          currentPanel: state.currentPanel - 1
        })
      } else {
        return state
      }

    case 'BOOKING_POST':
      console.log('BOOKING_POST')
      return state

    default:
      return state
  }
}

export default reducer
