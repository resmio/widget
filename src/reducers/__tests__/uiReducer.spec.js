import reducer from '../uiReducer'

import {
  UI_DATE_SELECTOR_CHANGE_STATE,
  UI_GUEST_SELECTOR_CHANGE_STATE,
  UI_TIME_SELECTOR_CHANGE_STATE,
  UI_PANEL_ADVANCE,
  UI_PANEL_REDUCE
} from '../../actions/uiActions'

import {
  AVAILABILITIES_FETCHING,
  AVAILABILITIES_FETCHING_ERROR,
  AVAILABILITIES_FETCHING_SUCCESS,
  DATE_SELECT,
  GUEST_SELECT,
  TIME_SELECT,
} from '../../actions/bookingActions'

import preloadedState from '../../preloadedState'

// We only need the ui part of the state
const state = preloadedState.ui

describe('ui reducer', () => {

  it('should set availabilities fetching to true', () => {
    const action = { type: AVAILABILITIES_FETCHING }
    expect(reducer(state, action)).toMatchSnapshot()
  })

  it('should set availabilities fetching errors', () => {
    const action = { type: AVAILABILITIES_FETCHING_ERROR }
    expect(reducer(state, action)).toMatchSnapshot()
  })

  it('should clear error and pending on availabilities fetching success', () => {
    const action = { type: AVAILABILITIES_FETCHING_SUCCESS }
    expect(reducer(state, action)).toMatchSnapshot()
  })

  describe('date selector', () => {

    describe('on date selector change state', () => {
      it('should expand itself and collapse everything else if itself is not expanded', () => {
        const testState = Object.assign(
          {}, state, {
            guestSelectorState: 'collapsed',
            dateSelectorState: 'semicollapsed',
            timeSelectorState  : 'collapsed'
          }
        )
        const action = { type: UI_DATE_SELECTOR_CHANGE_STATE }
        const expected = 'collapsed'
        const actual = reducer(testState, action)

        expect(actual.guestSelectorState).toEqual(expected)
        expect(actual.dateSelectorState).toEqual('expanded')
        expect(actual.timeSelectorState).toEqual(expected)
      })

      it('should semicollapse everything if itself is expanded', () => {
        const testState = Object.assign(
          {}, state, {
            guestSelectorState: 'collapsed',
            dateSelectorState: 'expanded',
            timeSelectorState  : 'collapsed'
          }
        )
        const action = { type: UI_DATE_SELECTOR_CHANGE_STATE }
        const expected = 'semicollapsed'
        const actual = reducer(testState, action)

        expect(actual.guestSelectorState).toEqual(expected)
        expect(actual.dateSelectorState).toEqual(expected)
        expect(actual.timeSelectorState).toEqual(expected)
      })
    })

    describe('on date select', () => {
      it('should expand itself and collapse everything else if itself is not expanded', () => {
        const testState = Object.assign(
          {}, state, {
            guestSelectorState: 'collapsed',
            dateSelectorState: 'semicollapsed',
            timeSelectorState  : 'collapsed'
          }
        )
        const action = { type: DATE_SELECT }
        const expected = 'collapsed'
        const actual = reducer(testState, action)

        expect(actual.guestSelectorState).toEqual(expected)
        expect(actual.dateSelectorState).toEqual('expanded')
        expect(actual.timeSelectorState).toEqual(expected)
      })

      it('should semicollapse everything if itself is expanded', () => {
        const testState = Object.assign(
          {}, state, {
            guestSelectorState: 'collapsed',
            dateSelectorState: 'expanded',
            timeSelectorState  : 'collapsed'
          }
        )
        const action = { type: DATE_SELECT }
        const expected = 'semicollapsed'
        const actual = reducer(testState, action)

        expect(actual.guestSelectorState).toEqual(expected)
        expect(actual.dateSelectorState).toEqual(expected)
        expect(actual.timeSelectorState).toEqual(expected)
      })
    })
  })

  describe('guest selector', () => {

    describe('on guest selector change state', () => {
      it('should expand itself and collapse everything else if itself is not expanded', () => {
        const testState = Object.assign(
          {}, state, {
            guestSelectorState: 'semicollapsed',
            dateSelectorState: 'semicollapsed',
            timeSelectorState  : 'semicollapsed'
          }
        )
        const action = { type: UI_GUEST_SELECTOR_CHANGE_STATE }
        const expected = 'collapsed'
        const actual = reducer(testState, action)

        expect(actual.guestSelectorState).toEqual('expanded')
        expect(actual.dateSelectorState).toEqual(expected)
        expect(actual.timeSelectorState).toEqual(expected)
      })

      it('should semicollapse everything if itself is expanded', () => {
        const testState = Object.assign(
          {}, state, {
            guestSelectorState: 'expanded',
            dateSelectorState: 'collapsed',
            timeSelectorState  : 'collapsed'
          }
        )
        const action = { type: UI_GUEST_SELECTOR_CHANGE_STATE }
        const expected = 'semicollapsed'
        const actual = reducer(testState, action)

        expect(actual.guestSelectorState).toEqual(expected)
        expect(actual.dateSelectorState).toEqual(expected)
        expect(actual.timeSelectorState).toEqual(expected)
      })
    })

    describe('on guest select', () => {
      it('should expand itself and collapse everything else if itself is not expanded', () => {
        const testState = Object.assign(
          {}, state, {
            guestSelectorState: 'semicollapsed',
            dateSelectorState: 'semicollapsed',
            timeSelectorState  : 'semicollapsed'
          }
        )
        const action = { type: GUEST_SELECT }
        const expected = 'collapsed'
        const actual = reducer(testState, action)

        expect(actual.dateSelectorState).toEqual(expected)
        expect(actual.guestSelectorState).toEqual('expanded')
        expect(actual.timeSelectorState).toEqual(expected)
      })

      it('should semicollapse everything if itself is expanded', () => {
        const testState = Object.assign(
          {}, state, {
            guestSelectorState: 'expanded',
            dateSelectorState: 'collapsed',
            timeSelectorState  : 'collapsed'
          }
        )
        const action = { type: GUEST_SELECT }
        const expected = 'semicollapsed'
        const actual = reducer(testState, action)

        expect(actual.guestSelectorState).toEqual(expected)
        expect(actual.dateSelectorState).toEqual(expected)
        expect(actual.timeSelectorState).toEqual(expected)
      })
    })

  })

  describe('time selector', () => {

    describe('on time selector change state', () => {
      it('should expand itself and collapse everything else if itself is not expanded', () => {
        const testState = Object.assign(
          {}, state, {
            guestSelectorState: 'semicollapsed',
            dateSelectorState: 'semicollapsed',
            timeSelectorState  : 'semicollapsed'
          }
        )
        const action = { type: UI_TIME_SELECTOR_CHANGE_STATE }
        const expected = 'collapsed'
        const actual = reducer(testState, action)

        expect(actual.guestSelectorState).toEqual(expected)
        expect(actual.dateSelectorState).toEqual(expected)
        expect(actual.timeSelectorState).toEqual('expanded')
      })

      it('should semicollapse everything if itself is expanded', () => {
        const testState = Object.assign(
          {}, state, {
            guestSelectorState: 'collapsed',
            dateSelectorState: 'collapsed',
            timeSelectorState  : 'expanded'
          }
        )
        const action = { type: UI_TIME_SELECTOR_CHANGE_STATE }
        const expected = 'semicollapsed'
        const actual = reducer(testState, action)

        expect(actual.guestSelectorState).toEqual(expected)
        expect(actual.dateSelectorState).toEqual(expected)
        expect(actual.timeSelectorState).toEqual(expected)
      })
    })

    describe('on time select', () => {
      it('should semicollapse everything', () => {
        const testState = Object.assign(
          {}, state, {
            guestSelectorState: 'collapsed',
            dateSelectorState: 'collapsed',
            timeSelectorState  : 'expanded'
          }
        )
        const action = { type: TIME_SELECT }
        const expected = 'semicollapsed'
        const actual = reducer(testState, action)

        expect(actual.guestSelectorState).toEqual(expected)
        expect(actual.dateSelectorState).toEqual(expected)
        expect(actual.timeSelectorState).toEqual(expected)
      })
    })

  })

  it('should advance the panel on panel advance', () => {
    const testState = Object.assign(
      {}, state, {
        currentPanel: 1
      }
    )
    const action = { type: UI_PANEL_ADVANCE }
    const expected = 2
    const actual = reducer(testState, action).currentPanel

    expect(actual).toEqual(expected)
  })

  it('should reduce the panel on panel reduce', () => {
    const testState = Object.assign(
      {}, state, {
        currentPanel: 2
      }
    )
    const action = { type: UI_PANEL_REDUCE }
    const expected = 1
    const actual = reducer(testState, action).currentPanel

    expect(actual).toEqual(expected)
  })


})
