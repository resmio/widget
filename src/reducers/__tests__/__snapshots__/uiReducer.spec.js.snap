exports[`ui reducer should clear error and pending on availabilities fetching success 1`] = `
Object {
  "availabilitiesError": false,
  "availabilitiesFetching": false,
  "currentPanel": 1,
  "dateSelectorState": "collapsed",
  "guestSelectorState": "collapsed",
  "timeSelectorState": "expanded",
}
`;

exports[`ui reducer should set availabilities fetching errors 1`] = `
Object {
  "availabilitiesError": true,
  "availabilitiesFetching": false,
  "currentPanel": 1,
  "dateSelectorState": "collapsed",
  "guestSelectorState": "collapsed",
  "timeSelectorState": "expanded",
}
`;

exports[`ui reducer should set availabilities fetching to true 1`] = `
Object {
  "availabilitiesError": false,
  "availabilitiesFetching": true,
  "currentPanel": 1,
  "dateSelectorState": "collapsed",
  "guestSelectorState": "collapsed",
  "timeSelectorState": "expanded",
}
`;
